### PR TITLE
Add new RPC methods for posts and reactions pallets

### PR DESCRIPTION
**Pull request by olehmell**

### CODE REVIEW

File: pallets/posts/rpc/runtime-api/src/lib.rs [hunk 1]
```diff
@@ -4,8 +4,8 @@ use codec::Codec;
 use sp_std::collections::btree_map::BTreeMap;
 use sp_std::vec::Vec;
 
-use pallet_posts::rpc::FlatPost;
-use pallet_utils::{PostId, SpaceId};
+use pallet_posts::rpc::{FlatPost, ExtFilter};
+use pallet_utils::{PostId, SpaceId };
```
Comment (F3Joule) [inline pallets/posts/rpc/runtime-api/src/lib.rs:8]:
```suggestion
use pallet_utils::{PostId, SpaceId};
```

File: pallets/posts/src/rpc.rs [hunk 2]
```diff
@@ -48,30 +85,40 @@ pub struct FlatPost<AccountId, BlockNumber> {
     pub score: i32,
 }
 
+#[derive(Encode, Decode, Ord, PartialOrd, Clone, Eq, PartialEq)]
+#[cfg_attr(feature = "std", derive(Serialize, Deserialize))]
+pub enum ExtFilter {
```
Comment (F3Joule) [inline pallets/posts/src/rpc.rs:90]: Maybe rename to more descriptive name? e.g. `PostExtensionFilter`

File: pallets/posts/rpc/runtime-api/src/lib.rs [hunk 3]
```diff
@@ -14,8 +14,10 @@ sp_api::decl_runtime_apis! {
     {
         fn get_next_post_id() -> PostId;
 
-        fn get_posts_by_ids(post_ids: Vec<PostId>) -> Vec<FlatPost<AccountId, BlockNumber>>;
-    
+        fn get_posts_by_ids(post_ids: Vec<PostId>, offset: u64, limit: u16) -> Vec<FlatPost<AccountId, BlockNumber>>;
+
+        fn get_public_posts(ext_filter: Vec<ExtFilter>, offset: u64, limit: u16) -> Vec<FlatPost<AccountId, BlockNumber>>;
```
Comment (F3Joule) [inline pallets/posts/rpc/runtime-api/src/lib.rs:19]:
```suggestion
        fn get_public_posts(ext_filters: Vec<ExtFilter>, offset: u64, limit: u16) -> Vec<FlatPost<AccountId, BlockNumber>>;
```

File: pallets/posts/rpc/src/lib.rs [hunk 4]
```diff
@@ -17,6 +17,17 @@ pub trait PostsApi<BlockHash, AccountId, BlockNumber> {
...
         post_ids: Vec<PostId>,
+        offset: u64,
+        limit: u16,
+    ) -> Result<Vec<FlatPost<AccountId, BlockNumber>>>;
+
+    #[rpc(name = "posts_getPublicPosts")]
+    fn get_public_posts(
+        &self,
+        at: Option<BlockHash>,
+        ext_filter: Vec<ExtFilter>,
```
Comment (F3Joule) [inline pallets/posts/rpc/src/lib.rs:28]:
```suggestion
        ext_filters: Vec<ExtFilter>,
```

File: pallets/posts/rpc/src/lib.rs [hunk 5]
```diff
@@ -104,12 +133,28 @@ where
...
+        let at = BlockId::hash(at.unwrap_or_else(|| self.client.info().best_hash));
+
+        let runtime_api_result = api.get_posts_by_ids(&at, post_ids, offset, limit);
+        runtime_api_result.map_err(map_rpc_error)
+    }
+
+    fn get_public_posts(
+        &self,
+        at: Option<<Block as BlockT>::Hash>,
+        ext_filter: Vec<ExtFilter>,
```
Comment (F3Joule) [inline pallets/posts/rpc/src/lib.rs:150]:
```suggestion
        ext_filters: Vec<ExtFilter>,
```

File: pallets/posts/src/rpc.rs [hunk 6]
```diff
@@ -155,33 +197,73 @@ impl<T: Trait> Module<T> {
...
+       Self::get_reply_ids_by_parent_ids(post_ids)
+           .into_iter()
+           .map(|(parent_id, reply_ids)| {
+                (parent_id, Self::get_posts_by_ids(reply_ids, offset, limit))
+            })
+           .collect::<BTreeMap<_, _>>()
+    }
+
+    pub fn get_public_posts(
+        ext_filter: Vec<ExtFilter>,
```
Comment (F3Joule) [inline pallets/posts/src/rpc.rs:240]:
```suggestion
        ext_filters: Vec<ExtFilter>,
```

File: runtime/src/lib.rs [hunk 7]
```diff
@@ -818,8 +818,12 @@ impl_runtime_apis! {
 
     impl posts_runtime_api::PostsApi<Block, AccountId, BlockNumber> for Runtime
     {
-		fn get_posts_by_ids(post_ids: Vec<PostId>) -> Vec<FlatPost<AccountId, BlockNumber>> {
-			Posts::get_posts_by_ids(post_ids)
+		fn get_posts_by_ids(post_ids: Vec<PostId>, offset: u64, limit: u16) -> Vec<FlatPost<AccountId, BlockNumber>> {
+			Posts::get_posts_by_ids(post_ids, offset, limit)
+		}
+
+		fn get_public_posts(ext_filter: Vec<ExtFilter>, offset: u64, limit: u16) -> Vec<FlatPost<AccountId, BlockNumber>> {
```
Comment (F3Joule) [inline runtime/src/lib.rs:825]:
```suggestion
		fn get_public_posts(ext_filters: Vec<ExtFilter>, offset: u64, limit: u16) -> Vec<FlatPost<AccountId, BlockNumber>> {
```

File: pallets/posts/rpc/src/lib.rs [hunk 8]
```diff
@@ -37,20 +48,38 @@ pub trait PostsApi<BlockHash, AccountId, BlockNumber> {
...
+    #[rpc(name = "posts_getRepliesByParentId")]
+    fn get_replies_by_parent_id(
+        &self,
+        at: Option<BlockHash>,
+        post_id: PostId,
+        offset: u64,
+        limit: u16,
+    ) -> Result<Vec<FlatPost<AccountId, BlockNumber>>>;
+
+    #[rpc(name = "posts_getRepliesByParentId")]
```
Comment (F3Joule) [inline pallets/posts/rpc/src/lib.rs:74]:
```suggestion
    #[rpc(name = "posts_getRepliesByParentIds")]
```

File: pallets/posts/rpc/src/lib.rs [hunk 9]
```diff
@@ -37,20 +48,38 @@ pub trait PostsApi<BlockHash, AccountId, BlockNumber> {
...
+    fn get_replies_by_parent_ids(
+        &self,
+        at: Option<BlockHash>,
+        post_ids: Vec<PostId>,
+        offset: u64,
+        limit: u16,
+    ) -> Result<BTreeMap<PostId,Vec<FlatPost<AccountId, BlockNumber>>>>;
```
Comment (F3Joule) [inline pallets/posts/rpc/src/lib.rs:81]:
```suggestion
    ) -> Result<BTreeMap<PostId, Vec<FlatPost<AccountId, BlockNumber>>>>;
```

File: pallets/posts/rpc/src/lib.rs [hunk 10]
```diff
@@ -141,23 +186,47 @@ where
...
+        runtime_api_result.map_err(map_rpc_error)
+    }
+
+    fn get_replies_by_parent_ids(
+        &self,
+        at: Option<<Block as BlockT>::Hash>,
+        post_ids: Vec<PostId>,
+        offset: u64,
+        limit: u16
+    ) -> Result<BTreeMap<PostId,Vec<FlatPost<AccountId, BlockNumber>>>> {
```
Comment (F3Joule) [inline pallets/posts/rpc/src/lib.rs:225]:
```suggestion
    ) -> Result<BTreeMap<PostId, Vec<FlatPost<AccountId, BlockNumber>>>> {
```

File: pallets/posts/src/rpc.rs [hunk 11]
```diff
@@ -1,43 +1,80 @@
 use codec::{Decode, Encode};
 #[cfg(feature = "std")]
 use serde::{Deserialize, Serialize};
-use sp_std::collections::btree_map::BTreeMap;
+use sp_std::collections::{ btree_map::BTreeMap, btree_set::BTreeSet };
```
Comment (F3Joule) [inline pallets/posts/src/rpc.rs:4]:
```suggestion
use sp_std::collections::{btree_map::BTreeMap, btree_set::BTreeSet};
```

**Review by F3Joule — CHANGES_REQUESTED** [review]

File: pallets/reactions/src/rpc.rs [hunk 12]
```diff
@@ -73,22 +72,26 @@ impl<T: Trait> Module<T> {
         reactions
     }
 
-    pub fn get_reactions_by_account_and_post_ids(
-        account: T::AccountId,
+    pub fn get_reactions_by_post_ids_and_responder(
         post_ids: Vec<PostId>,
-    ) -> BTreeMap<PostId, FlatReaction<T::AccountId, T::BlockNumber>> {
+        reactor: T::AccountId,
```
Comment (F3Joule) [inline pallets/reactions/src/rpc.rs:77]: So responder or reactor?

**Review by F3Joule — CHANGES_REQUESTED** [review]

**Review by siman — CHANGES_REQUESTED** [review]

**Pull request merged**